### PR TITLE
Rework UBOS implementation

### PR DESCRIPTION
**Pull request by jfroy**
This patch reworks the UBOS router implementation to use dnsmasq forwarding. This fixes domain-based traffic management rules, which rely on ipset config entries managed by UBOS.

The new implementation also removes container (Podman) support, since UBOS 2.x and later do not use it and 1.x products are being migrated.

I don't know the reason why the UBOS implementation was using firewall rules instead of dnsmasq forwarding, but running this fork on my lab environment has been working quite well.

I'm happy to discuss the patch and make changes if there's interest in merging it.

### CODE REVIEW

File: router/ubios/setup.go [hunk 1]
```diff
@@ -1,159 +1,75 @@
...
-		if flds[2] != "/" && flds[2] != "/init.scope" {
-			return true, nil
-		}
-	}
-	return false, nil
+func (r *Router) Configure(c *config.Config) error {
+	c.Listens = []string{net.JoinHostPort("localhost", r.ListenPort)}
+	r.ClientReporting = c.ReportClientInfo
+	// Domain-based traffic management rules depend on dnsmasq handling requests. So forcefully disable caching.
+	c.CacheSize = "0"
```
Comment (rs) [inline router/ubios/setup.go:34]: Keeping dnsmasq cache in front of the CLI would remove the ability to use conditional profiles, as you need to maintain a segmented cache per profile.

**Comment by rs** [comment]
This implementation would break conditional profiles support unless you disable dnsmasq caching. It would also break device name discovery as we use the local dnsmasq to query for PTR in order to get the names you set in the UI and we would risk loops in this setup.

Removing podman support now would likely break people who did not upgrade yet or devices stuck in 1.x for some reason.

File: router/ubios/setup.go (hunk 1 above)
Comment (jfroy) [reply router/ubios/setup.go:34]: Ah yes indeed. I bet caching in dnsmasq can be disabled in the config, and then nextdns can handle caching.

**Comment by jfroy** [comment]
> This implementation would break conditional profiles support unless you disable dnsmasq caching. It would also break device name discovery as we use the local dnsmasq to query for PTR in order to get the names you set in the UI and we would risk loops in this setup.

Won't device name be provided via the `add-mac` and `add-subnet=32,128` options?

> Removing podman support now would likely break people who did not upgrade yet or devices stuck in 1.x for some reason.

That's fair. I'll see if I can integrate it back. It seemed like a good simplification but I don't want to force people to upgrade (especially given that it's still in EA). Maybe once UI deprecates 1.x (if they ever do).

**Comment by rs** [comment]
The add-mac and add-subnet provides the IP and the mac of the client, but then we need to convert it to a name. Since there is no API on the UDM to translate an IP or a MAC to a name, we use the local DNS for this.

**Comment by jfroy** [comment]
> The add-mac and add-subnet provides the IP and the mac of the client, but then we need to convert it to a name. Since there is no API on the UDM to translate an IP or a MAC to a name, we use the local DNS for this.

Would it be good enough to scrape that from the dnsmasq lease file (`/data/udapi-config/dnsmasq.lease`)?

**Comment by rs** [comment]
Those won't get the personalized name set in the UI.

**Comment by jfroy** [comment]
> Those won't get the personalized name set in the UI.

Those get scraped by the existing `Ubios` resolver that reads the mongo database. I assume resolvers are ordered so those take precedence over the `DHCP` resolver.

**Comment by rs** [comment]
Actually, I forgot that we had specific UDM discovery support here https://github.com/nextdns/nextdns/blob/master/discovery/ubios_linux.go. The caching issue with conditional profile still stand though. The cache-size should be set to 0.

**Comment by jfroy** [comment]
Yep, I'm reworking the patch to incorporate all the feedback.

**Comment by jimmystewpot** [comment]
One aspect of this proposal which may also need to be considered, in the existing ubios implementation it runs the iptables and ip6tables commands. This hard codes the configuration so that it only works on a [single br0 interface](https://www.ndis.gov.au/providers/housing-and-living-supports-and-services/housing).  To make nextdns work on multi-network configurations the network admin needs to add each additional bridge manually like so

```
ip6tables -t nat -I PREROUTING 3 -j NEXTDNS -m set --match-set UBIOS6ADDRv6_br2 dst
ip6tables -t nat -I PREROUTING 4 -j NEXTDNS -m set --match-set UBIOS6ADDRv6_br3 dst
```

With dnsmasq the administrator no longer needs to add each network manually which is often overwritten during updates without prior planning.

**Comment by jfroy** [comment]
Dealing with an vexing issue -- dnsmasq refuses to start if there are two or more `cache-size` directives in its config file. I'm researching an alternative to control caching.

**Comment by rs** [comment]
I merged this too soon. Please reopen a new PR when you are ready with the last changes.

**Comment by jfroy** [comment]
> I merged this too soon. Please reopen a new PR when you are ready with the last changes.

Will do. I was quite confused by that merge!

**Comment by rs** [comment]
@jfroy sorry about that, I was  too impatient :)

Regarding dnsmasq config, one solution would be to make CLI listen on 53 and dnsmasq listen on 5353 or something and setup `discovery-dns` to `127.0.0.1:5353`.

**Comment by jfroy** [comment]
> @jfroy sorry about that, I was  too impatient :)
> 
> 
> 
> Regarding dnsmasq config, one solution would be to make CLI listen on 53 and dnsmasq listen on 5353 or something and setup `discovery-dns` to `127.0.0.1:5353`.

I'll look into that. But generally speaking dnsmasq needs to see every request and it's result so it can update the ipsets listed in its config so that domain-based traffic rules (which allow or deny traffic using a domain list) work. I could implement a parser to read those lines from the dnsmasq config and have the CLI update the ipsets, but that's just chasing every feature in the network app that depends on dnsmasq integration.

**Comment by rs** [comment]
I wasn't aware of such feature and it feels redundant with NextDNS (and the wrong way of implementing such filtering if you ask me :). I would probably not obsess about supporting it if it makes the whole system simpler and more robust.

**Comment by rs** [comment]
Another solution would be to support a tap DNS address where the CLI would send all queries without listening for answers.

**Comment by jfroy** [comment]
> I wasn't aware of such feature and it feels redundant with NextDNS (and the wrong way of implementing such filtering if you ask me :). I would probably not obsess about supporting it if it makes the whole system simpler and more robust.

It's for sure not super good and indeed DNS blocking is much better (why I use NextDNS). But, and this sucks so much, some IoT devices will refuse to work if they cannot resolve DNS entries, but work if they resolve but the TCP connections get dropped by a firewall. Possibly NextDNS could add a feature to block certain domains by returning an unroutable address instead of no result.

**Comment by rs** [comment]
NextDNS is blocking by returning 0.0.0.0.

**Comment by rs** [comment]
@jfroy are you still planning to submit an updated PR for this?

**Comment by jimmystewpot** [comment]
@jfroy 
> Dealing with an vexing issue -- dnsmasq refuses to start if there are two or more cache-size directives in its config file. I'm researching an alternative to control caching. 

a possible hack is to set the max-cache-ttl to 1

**Comment by jfroy** [comment]
> @jfroy are you still planning to submit an updated PR for this?

Yes, but I haven't had time to do the research to solve the cache issue.

**Comment by jfroy** [comment]
> @jfroy 
> 
> > Dealing with an vexing issue -- dnsmasq refuses to start if there are two or more cache-size directives in its config file. I'm researching an alternative to control caching. 
> 
> 
> 
> a possible hack is to set the max-cache-ttl to 1

Appealing, worth a try.

**Comment by lps-rocks** [comment]
Any updates to this effort? I've had to disable nextdns on my UDM:SE so i can use the local hostnames.

**Comment by rs** [comment]
I tested `max-cache-ttl` set to `0` and it works. I'll push a change with this trick.

**Comment by jfroy** [comment]
Ah nice. I haven't had time to work on this PR at all recently.

**Pull request merged**